core/src/main/java/org/mskcc/cbio/portal/scripts/ImportResourceData.java

```diff
@@ -0,0 +1,555 @@
+                    throw new RuntimeException(
+                            "Something has gone wrong. Patient " + patientId + " already has resource data loaded.");
+                }
+                // if patient is duplicated, abort as well in this case:
+                if (!patientIds.add(patientId)) {
+                    throw new RuntimeException("Error. Patient " + patientId + " found to be duplicated in your file.");
+                }
+            }
+
+            if (patient != null) {
+                // in all cases (SAMPLE, PATIENT, or STUDY) this can be expected, so
+                // just fetch it:
+                internalPatientId = patient.getInternalId();
+            } else {
+                // in case of PATIENT data import and patient == null :
+                if (getResourceType() == ResourceType.PATIENT) {
+                    // not finding the patient it unexpected (as SAMPLE data import should always
+                    // precede it), but
+                    // can happen when this patient does not have any samples for example. In any
+                    // case, warn about it:
+                    ProgressMonitor.logWarning("Patient " + patientId
+                            + " being added for the first time. Apparently this patient was not in the samples file, or the samples file is not yet loaded (should be loaded before this one)");
+                }
+
+                patient = new Patient(cancerStudy, patientId);
+                internalPatientId = DaoPatient.addPatient(patient);
+            }
+        }
+        return internalPatientId;
+    }
+
+    private int addSampleToDatabase(String sampleId, String[] fields, Map<String, Integer> headerIndexMap)
+            throws Exception {
+        int sampleTypeIndex = findSampleTypeColumn(headerIndexMap);
+        String sampleTypeStr = (sampleTypeIndex != -1) ? fields[sampleTypeIndex] : null;
+        if (sampleTypeStr != null) {
+            // want to match Sample.Type enum names
+            sampleTypeStr = sampleTypeStr.trim().toUpperCase().replaceAll(" ", "_");
+        }
+        Sample.Type sampleType = Sample.Type.has(sampleTypeStr) ? Sample.Type.valueOf(sampleTypeStr) : null;
+
+        int internalSampleId = -1;
+        if (validSampleId(sampleId) && !StableIdUtil.isNormal(sampleId)) {
+            // want to try and capture normal sample types based on value for SAMPLE_TYPE
+            // if present in resource data
+            if (sampleType != null && sampleType.isNormal()) {
+                return internalSampleId;
+            }
+            String stablePatientId = getStablePatientId(sampleId, fields, headerIndexMap);
+            if (validPatientId(stablePatientId)) {
+                Patient patient = DaoPatient.getPatientByCancerStudyAndPatientId(cancerStudy.getInternalId(),
+                        stablePatientId);
+                if (patient == null) {
```

Comment:
I think `addSampleToDatabase` should be executed only when patient is added. I prefer passing patient object in the method parameter. this would reduce the number of queries made to database to fetch patient(`getPatientByCancerStudyAndPatientId`)